rust/xaynet-analytics/src/database/instance.rs

```
@@ -0,0 +1,116 @@
+            .build()
+            .find_all_vec(&transaction)
+        {
+            Ok(bytes) => bytes,
+            Err(e) => return Err(e),
+        };
+
+        // TODO: not sure how to proceed to parse [u8] using the collection schema. didn't find examples in Isar
+        unimplemented!()
+    }
+
+    pub fn put(&self, collection_name: &str, object: &[u8]) -> Result<(), IsarError> {
+        let transaction = match self.instance().begin_txn(false) {
+            Ok(transaction) => transaction,
+            Err(e) => return Err(e),
+        };
+        let collection = match self.get_collection(collection_name) {
+            Ok(collection) => collection,
+            Err(e) => return Err(e),
+        };
+        match collection.put(&transaction, None, object) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn get_object_builder(&self, collection_name: &str) -> Result<ObjectBuilder, IsarError> {
+        match self.get_collection(collection_name) {
+            Ok(collection) => Ok(collection.get_object_builder()),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn get_schema() -> Schema {
+        let mut schema = Schema::new();
+        schema
+            .add_collection(get_collection_schema(
+                Self::ANALYTICS_EVENT_NAME,
+                AnalyticsEvent::into_field_properties(),
+            ))
+            .ok();
+        // here add more collections schemas
+        schema
+    }
+
+    fn get_collection(&self, collection_name: &str) -> Result<&IsarCollection, IsarError> {
+        match self.instance().get_collection_by_name(collection_name) {
+            Some(collection) => Ok(collection),
+            None => Err(IsarError::IllegalArg {
+                message: "wrong collection name".to_string(),
+            }),
+        }
+    }
+}
+
+fn get_collection_schema(
+    name: &str,
+    field_properties: IntoIter<FieldProperty>,
+) -> CollectionSchema {
```

We should propagate the error.

---

done.. i think